### PR TITLE
#73 Add property to executescheduler on startup

### PR DESCRIPTION
**Pull request by thwint**
Added property executeOnStartup

### CODE REVIEW

File: gateleen-scheduler/src/main/java/org/swisspush/gateleen/scheduler/Scheduler.java [hunk 1]
```diff
@@ -92,6 +95,15 @@ public void start() {
                     }
                 });
             }
+            if(!executed && executeOnStartup ){
+                executed = true;
+                vertx.setTimer((randomOffset + 1) * 1000, new Handler<Long>() {
```
Comment (lbovet) [inline gateleen-scheduler/src/main/java/org/swisspush/gateleen/scheduler/Scheduler.java]: We should not have randomOffset here. trigger should be called directly

File: gateleen-test/src/test/java/org/swisspush/gateleen/scheduler/SchedulerTest.java [hunk 2]
```diff
@@ -169,6 +169,108 @@ public void testSchedulerWithRandomOffset(TestContext context) throws Exception
...
+        String schedulerName = "scheduler-executeOnStartup-test";
+        String resourceName = "schedulerWithExecuteInStartup";
+        String cronExpression = "* * 1/1 * * ?";
+
+        // scheduler config
+        String testSchedulers = "{\n"
+                + "  \"schedulers\": {\n"
+                + "    \"" + schedulerName + "\": {\n"
+                + "      \"cronExpression\": \"" + cronExpression + "\",\n"
+                + "      \"executeOnStartup\": " + true + ",\n"
```
Comment (lbovet) [inline gateleen-test/src/test/java/org/swisspush/gateleen/scheduler/SchedulerTest.java]: You can write true in the string directly

File: gateleen-test/src/test/java/org/swisspush/gateleen/scheduler/SchedulerTest.java [hunk 3]
```diff
@@ -169,6 +169,108 @@ public void testSchedulerWithRandomOffset(TestContext context) throws Exception
...
+        // wait for the scheduler to be created
+        await().atMost(10, SECONDS).until( () -> {
+            if ( schedulerResourceManager.getSchedulers() != null ) {
+                return schedulerResourceManager.getSchedulers().stream().filter(s -> s.getName().equals(schedulerName)).findFirst().get();
+            }
+            return null;
+        }, is(notNullValue()));
+
+        // get the scheduler and the randomOffset
+        Scheduler scheduler = schedulerResourceManager.getSchedulers().stream().filter(s -> s.getName().equals(schedulerName)).findFirst().get();
```
Comment (lbovet) [inline gateleen-test/src/test/java/org/swisspush/gateleen/scheduler/SchedulerTest.java]: Where is then "scheduler" used?

File: gateleen-test/src/test/java/org/swisspush/gateleen/scheduler/SchedulerTest.java [hunk 4]
```diff
@@ -169,6 +169,108 @@ public void testSchedulerWithRandomOffset(TestContext context) throws Exception
...
+    @Test
+    public void testSchedulerWithExecuteOnStartupTrue(TestContext context) throws Exception {
+        Async async = context.async();
+        delete();
+
+        // prepare the settings
+        String content = String.valueOf(System.currentTimeMillis());
+        String schedulerName = "scheduler-executeOnStartup-test";
+        String resourceName = "schedulerWithExecuteInStartup";
+        String cronExpression = "* * 1/1 * * ?";
```
Comment (lbovet) [inline gateleen-test/src/test/java/org/swisspush/gateleen/scheduler/SchedulerTest.java]: How often does this cron expression runs?

File: gateleen-test/src/test/java/org/swisspush/gateleen/scheduler/SchedulerTest.java [hunk 5]
```diff
@@ -169,6 +169,108 @@ public void testSchedulerWithRandomOffset(TestContext context) throws Exception
...
+                return schedulerResourceManager.getSchedulers().stream().filter(s -> s.getName().equals(schedulerName)).findFirst().get();
+            }
+            return null;
+        }, is(notNullValue()));
+
+        // get the scheduler and the randomOffset
+        Scheduler scheduler = schedulerResourceManager.getSchedulers().stream().filter(s -> s.getName().equals(schedulerName)).findFirst().get();
+
+        // wait for the vertx timer to fire the first time
+        await().atMost(10, SECONDS).until( () -> jedis.get("schedulers:" + schedulerName), is(notNullValue()));
```
Comment (lbovet) [inline gateleen-test/src/test/java/org/swisspush/gateleen/scheduler/SchedulerTest.java]: Here you test that the scheduler has run...

File: gateleen-test/src/test/java/org/swisspush/gateleen/scheduler/SchedulerTest.java [hunk 6]
```diff
@@ -169,6 +169,108 @@ public void testSchedulerWithRandomOffset(TestContext context) throws Exception
...
+
+        // Test if first run
+        await().atMost(30, SECONDS).until(() -> get("/tests/" + SERVER_NAME + "/" +resourceName ).then().extract().statusCode(), equalTo(404));
```
Comment (lbovet) [inline gateleen-test/src/test/java/org/swisspush/gateleen/scheduler/SchedulerTest.java]: ...and here you test that it had no effect. I don't see that as OK.

File: gateleen-scheduler/src/main/java/org/swisspush/gateleen/scheduler/Scheduler.java [hunk 7]
```diff
@@ -92,6 +95,15 @@ public void start() {
                     }
                 });
             }
+            if(!executed && executeOnStartup ){
```
Comment (lbovet) [inline gateleen-scheduler/src/main/java/org/swisspush/gateleen/scheduler/Scheduler.java]: Why is this in the setPeriodic? We should just call trigger() inside start, no?

**Comment by thwint** [comment]
Thank you for this review. You are absolutely right. I will correct this. 
Only one question concerning the tests. 
We don't want to test if the schedulers are executed periodically. We already do this within other tests. But we want to ensure that in one case the scheduler is triggered at startup and in the other case it is not.
Do you have a proposal how to test that it is ok?

**Pull request merged**